go/app/main.go

```diff
@@ -70,33 +92,75 @@ func addItem(c echo.Context) error {
 	return c.JSON(http.StatusOK, res)
 }
 
-// saveItem writes the item information into the JSON file.
+// saveItem writes the item information into the database.
 func saveItem(name, category, fileName string) error {
 
-	item := Item{Name: name, Category: category, ImageName: fileName}
-
-	items, err := readItems(ItemsPath)
+	dbCon, err := connectDB(DBPath)
 	if err != nil {
-		err := fmt.Errorf("error while reading or unmarshaling file: %w", err)
 		return err
 	}
+	defer dbCon.Close()
 
-	items.Items = append(items.Items, item)
+	// Transaction starts.
+	tx, err := dbCon.Begin()
+	if err != nil {
+		err = fmt.Errorf("error while beginning transaction: %s", err)
+		return err
+	}
 
-	newData, err := json.Marshal(items)
+	categoryId, err := searchCategoryId(category, tx)
 	if err != nil {
-		err := fmt.Errorf("error while marshaling file: %w", err)
+		tx.Rollback()
+		err = fmt.Errorf("error while searching for category id: %s", err)
 		return err
 	}
 
-	if err = os.WriteFile(ItemsPath, newData, 0644); err != nil {
-		err := fmt.Errorf("error while writing file: %w", err)
+	insertItem := "INSERT INTO items (name, image_name, category_id) VALUES (?, ?, ?)"
+	_, err = tx.Exec(insertItem, name, fileName, categoryId)
+	if err != nil {
+		tx.Rollback()
+		err = fmt.Errorf("error while adding a new item: %s", err)
 		return err
+	} else {
+		if err := tx.Commit(); err != nil {
+			err = fmt.Errorf("error while commiting transaction: %s", err)
+			return err
+		}
 	}
 
 	return nil
 }
 
+// searchCategoryId look for category ID, if it doesn't exist, register a new category.
+func searchCategoryId(category string, tx *sql.Tx) (int, error) {
+	var categoryId int
+
+	searchForKey := "SELECT id FROM categories WHERE name = ?"
```

Comment:
imo: コンパイル時に確定する文字列なので、 `const` で宣言してあげると良さそうです 👍 
```suggestion
	const searchForKey = "SELECT id FROM categories WHERE name = ?"
```

Reply:
ありがとうございます！

うっかり変えてしまってエラーになったりしないように……という意図で合っていますか？
ここ以外のクエリも定数宣言しました。
https://github.com/70ru5/mercari-build-training/pull/3/commits/8d95e0a7f20cecffb95cccaad261bdfb488b5877

Reply:
前提として、この話はオプショナルの話なのでスルーしていただいて大丈夫です
興味があれば読んでください :pray:

> うっかり変えてしまってエラーになったりしないように……という意図で合っていますか？

本当に細かい話になりますが、コンパイラ最適化によって実行ファイルの実行速度を高速化するためです！
コンパイラとは、私たちが書いたGoのコードを、機械語で記述された実行ファイルにコンパイル(変換)してくれるソフトウェアのことで、 `$ go build .` を実行するとGoコンパイラがGoのコードを実行ファイルにコンパイルしてくれています

このコンパイルされた高速に実行できるように、コンパイラは静的な情報(今回のように、定数で宣言された変化しない情報)をもとに最適化を行います
そのため、可能なら私たちプログラマは、可能な限り定数にできるものは定数で書いた方が高速に実行ファイルを実行することができますし、そうする方が良いです

詳しくはこの辺りの記事が分かりやすいと思いますが、非本質的な話なのでスルーしてもらっても大丈夫です 🙆‍♂️ 
https://qiita.com/uint256_t/items/a0d050894927b618daa3#%E5%AE%9A%E6%95%B0%E7%95%B3%E3%81%BF%E8%BE%BC%E3%81%BF

Reply:
詳しく説明してくださってありがとうございます！
まったく知らなかったことで、とても勉強になりました。